File: python/349.intersection-of-two-arrays.py
from typing import List


class Solution:
    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        """Use hashmap.

        Iterate first list and add to hashmap. Then iterate second list,
        if element is already in hashmap, add to result.
        """

        cache = {}
        for num in nums1:
            cache[num] = 1
        res = []
        for num in nums2:
            if num in cache and cache[num] == 1:
                res.append(num)
                cache[num] = 2

        return res

File: cpp/349.intersection-of-two-arrays.cpp
class Solution {
public:
  vector<int> intersection(vector<int> &nums1, vector<int> &nums2) {
    unordered_set<int> set;
    vector<int> res;

    for (auto num : nums1) {
      set.insert(num);
    }

    for (auto num : nums2) {
      if (set.find(num) != set.end()) {
        res.push_back(num);
        set.erase(num);
      }
    }

    return res;
  }
};